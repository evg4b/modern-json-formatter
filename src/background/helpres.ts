import { JqParams, Message, TokenizeParams } from '@core/background';

interface Types {
  jq: JqParams;
  tokenize: TokenizeParams;
}

export const is = <T extends Message['action']>(message: object, type: T): message is Types[T] => {
  return 'action' in message && message.action == type;
};
