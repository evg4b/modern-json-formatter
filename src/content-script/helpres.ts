export const throws = (value?: string): never => {
  throw new Error(value ?? 'Unexpected value');
};

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-parameters
export const query = <T extends HTMLElement = HTMLElement, D extends HTMLElement = HTMLElement>(
  value: D,
  selector: string
): T => value.querySelector<T>(selector) ?? throws(`Element ${selector} not found`);

const types: string[] = ['raw', 'query', 'formatted'] satisfies TabType[];

export function assetTabType(s?: string | null): asserts s is TabType {
  if (s && !types.includes(s)) {
    throw new Error(`Invalid tab type '${s}'`);
  }
}
