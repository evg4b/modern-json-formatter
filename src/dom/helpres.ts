export const element = (options?: { content?: string, class?: string }) => {
  const span = document.createElement('span');
  if (options?.content) {
    span.appendChild(document.createTextNode(options.content));
  }
  if (options?.class) {
    span.className = options.class;
  }

  return span;
};

export const isValueExpandable = (value: ParsedJSON): boolean =>
  value.type === 'object' && !!value.properties.length
  || value.type === 'array' && !!value.items.length;

export const isToggleElement = (element: EventTarget | null): element is HTMLElement => {
  return element instanceof HTMLElement && element.classList.contains('toggle');
};
