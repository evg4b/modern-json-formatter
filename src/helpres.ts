export const throws = (value?: string): never => {
  throw new Error(value ?? 'Unexpected value');
};

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-parameters
export const query = <T extends HTMLElement = HTMLElement, D extends HTMLElement = HTMLElement>(
  value: D,
  selector: string
): T => value.querySelector<T>(selector) ?? throws(`Element ${selector} not found`);
