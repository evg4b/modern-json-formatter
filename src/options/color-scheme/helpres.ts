import { ColorScheme, Preset } from './models';

function compileColorScheme(scheme: ColorScheme | undefined) {
  if (!scheme) {
    return null;
  }

  return Object.entries(scheme)
    .map((prop) => prop.join(':'))
    .join(';');
}

export const compilePreset = (preset: Preset): string => {
  const light = compileColorScheme(preset.light);
  const dark = compileColorScheme(preset.dark);

  if (!light && !dark) {
    console.warn('Preset is empty');

    return '';
  }

  if (light && dark) {
    return [
      `@media(prefers-color-scheme:dark){:host{${ dark }}}`,
      `@media(prefers-color-scheme:light){ :host{${ light }}}`,
    ].join('\n');
  }

  return `:host{${ light ?? dark }}`;
};
