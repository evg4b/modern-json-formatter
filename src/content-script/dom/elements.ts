import { element } from './helpres';

export const toggle = () => element({ class: 'toggle' });
export const ellipsis = () => element({ class: 'ellipsis' });
export const comma = () => element({ content: ',', class: 'comma' });
export const colon = () => element({ content: ':', class: 'colon' });
export const bracket = {
  open: () => element({ content: '{', class: 'bracket bracket-open' }),
  close: () => element({ content: '}', class: 'bracket bracket-close' }),
};

export const squareBracket = {
  open: () => element({ content: '[', class: 'bracket square-bracket-open' }),
  close: () => element({ content: ']', class: 'bracket square-bracket-close' }),
};

export const propertiesCount = (count: number) =>
  element({
    content: `// ${count} propert${count === 1 ? 'y' : 'ies'}`,
    class: 'properties-count',
  });

export const itemsCount = (count: number) =>
  element({
    content: `// ${count} item${count === 1 ? '' : 's'}`,
    class: 'items-count',
  });
