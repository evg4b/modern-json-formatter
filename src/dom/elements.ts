import { element } from './helpres';

export const toggle = () => element({ class: 'toggle' });
export const ellipsis = () => element({ class: 'ellipsis' });
export const comma = () => element({ content: ',', class: 'comma' });
export const colon = () => element({ content: ':', class: 'colon' });
export const bracket = {
  open: () => element({ content: '{', class: 'bracket bracket-open' }),
  close: () => element({ content: '}', class: 'bracket bracket-close' }),
};
export const squareBracket = {
  open: () => element({ content: '[', class: 'bracket square-bracket-open' }),
  close: () => element({ content: ']', class: 'bracket square-bracket-close' }),
};
