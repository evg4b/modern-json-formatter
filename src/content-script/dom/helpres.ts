import { ArrayNode, ObjectNode, TokenNode } from '@packages/tokenizer';
import { itemsCount, propertiesCount } from './elements';

export const element = (options?: { content?: string; class?: string; block?: boolean }) => {
  const span = document.createElement(!options?.block ? 'span' : 'div');
  if (options?.content) {
    span.appendChild(document.createTextNode(options.content));
  }
  if (options?.class) {
    span.className = options.class;
  }

  return span;
};

interface Options<K extends keyof HTMLElementTagNameMap = 'span'> {
  element: K;
  content?: string;
  class?: string | string[];
  children?: HTMLElement[] | HTMLElement;
  attributes?: { [key: string]: string };
}

export const createElement = <K extends keyof HTMLElementTagNameMap>(options: Options<K>): HTMLElementTagNameMap[K] => {
  const element = document.createElement(options.element ?? 'span');
  if (options.content) {
    element.appendChild(document.createTextNode(options.content));
  }
  if (options.class) {
    const classList = [options.class].flat();
    element.classList.add(...classList);
  }
  if (options.children) {
    const children = [options.children].flat();
    element.append(...children);
  }
  if (options.attributes) {
    Object.entries(options.attributes).forEach(([key, value]) => {
      element.setAttribute(key, value);
    });
  }

  return element;
};

export const isValueExpandable = (value: TokenNode): value is ObjectNode | ArrayNode =>
  (value.type === 'object' && !!value.properties.length) || (value.type === 'array' && !!value.items.length);

export const isToggleElement = (element: EventTarget | null): element is HTMLElement => {
  return element instanceof HTMLElement && element.classList.contains('toggle');
};

export const buildInfoNode = (value: TokenNode): HTMLSpanElement | null => {
  if (isValueExpandable(value)) {
    switch (value.type) {
      case 'array':
        return itemsCount(value.items.length);
      case 'object':
        return propertiesCount(value.properties.length);
      default:
        throw new Error('Unknown type');
    }
  }

  return null;
};
