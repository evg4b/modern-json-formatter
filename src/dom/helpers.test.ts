import { tArray, tBool, tNull, tNumber, tObject, tProperty, tString } from '../../testing';
import { colon, toggle } from './elements';
import { element, isToggleElement, isValueExpandable } from './helpres';

describe('dom helpers', () => {
  describe('isValueExpandable', () => {
    describe('should return true for', () => {
      const testCases = [
        {
          name: 'not empty object',
          value: tObject(
            tProperty('demo', tBool(true)),
          ),
        },
        {
          name: 'not empty array',
          value: tArray(
            tString('demo'),
          ),
        },
      ];
      test.each(testCases)('$name', ({ value }) => {
        expect(isValueExpandable(value)).toBe(true);
      });
    });

    describe('should return false for', () => {
      const testCases = [
        { name: 'null', value: tNull() },
        { name: 'boolean', value: tBool(true) },
        { name: 'string', value: tString('demo') },
        { name: 'number', value: tNumber('123') },
        { name: 'empty object', value: tObject() },
        { name: 'empty array', value: tArray() },
      ];
      test.each(testCases)('$name', ({ value }) => {
        expect(isValueExpandable(value)).toBe(false);
      });
    });
  });

  describe('isToggleElement', () => {
    test('should return true for toggle element', () => {
      expect(isToggleElement(toggle())).toBe(true);
    });
    test('should return false for non-toggle element', () => {
      expect(isToggleElement(colon())).toBe(false);
    });
  });

  describe('element', () => {
    test('should create element with content', () => {
      expect(element({ content: 'demo' })).toMatchSnapshot();
    });

    test('should create element with class', () => {
      expect(element({ class: 'demo' })).toMatchSnapshot();
    });

    test('should create element with content and class', () => {
      expect(element({ content: 'demo', class: 'demo' })).toMatchSnapshot();
    });
  });
});
