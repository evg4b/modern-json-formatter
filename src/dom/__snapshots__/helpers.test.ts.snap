// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dom helpers element should create element with class 1`] = `
<span
  class="demo"
/>
`;

exports[`dom helpers element should create element with content 1`] = `
<span>
  demo
</span>
`;

exports[`dom helpers element should create element with content and class 1`] = `
<span
  class="demo"
>
  demo
</span>
`;
